lib/spack/spack/package_base.py

```diff
@@ -194,18 +194,29 @@ class WindowsRPathMeta(object):
     they would a genuine RPATH, i.e. adding directories that contain
     runtime library dependencies"""
 
-    def add_search_paths(self, *path):
+    def add_search_paths(self, *paths):
         """Add additional rpaths that are not implicitly included in the search
         scheme
         """
-        self.win_rpath.include_additional_link_paths(*path)
+        self.win_rpath.include_additional_link_paths(*paths)
+
+    def add_internal_links(self, *paths):
+        """Add additional rpaths internal to a package, i.e.
+        linking pkg.prefix.bin to pkg.prefix.lib.site-packages"""
+        self.win_rpath.add_internal_links(*paths)
+
+    def win_setup_rpath(self):
+        """This method should be overridden by packages needing bespoke RPATH
+        support. No-op otherwise"""
+        pass
```

Comment:
(1) 

I assume this would never be more in-depth than supplying a set of additional directories that should be RPATHed. I think the interface can be narrowed then so that packages may define a function that returns a list of directories (rather than being an open-ended function that can do whatever it wants).

Moreover, I think we can get away with removing `package_base.add_internal_links` and just do:

```
    def windows_establish_runtime_linkage(self):
         """Establish RPATH on Windows

         Performs symlinking to incorporate rpath dependencies to Windows runtime search paths
         """
         if is_windows:
             self.win_rpath.add_internal_links(self.win_setup_rpath())
             self.win_rpath.establish_link()
```

Reply:
I'm fine with these changes, however we may want to consider the case where a package's `win_setup_rpath` method may also want to invoke the `WindowsRPathMeta.add_search_paths` (which extends the directories that the bin/lib dirs link against) as opposed or in addition to the internal links. Ostensibly this new refactored version could call that method and return a set of directories for internal linking but that feels a bit convoluted.

Reply:
Do we need users to explicitly decide whether to invoke `add_internal_links` or `add_search_paths`? Can we have (from the user's perspective) a single function for this, and then examine the paths to decide whether they are internal or not?

Reply:
Maybe internal isn't the best term (my fault for using it) to articulate what I was considering here. Essentially, I wanted the `add_search_paths` to behave as appending a path to the `-rpath` arg given to ld would, whereas the  `add_internal_links` was intended to be more along the lines of adding a binary/library to the manifest of things that need linking, and it would then be linked against everything in the RPath args. In practice I deviated from this slightly, `add_internal_links` does not link against its package deps derived rpath, and bin/lib dirs do not link against each other. This was done to avoid over linking and causing conflicts or issues. I can include this extra behavior very easily if we want.

To answer your question more directly, I think parsing the path to determine if it's inside of the package isn't really solving the issue here, as the `add_search_path` method could point to something inside the package prefix, but that isn't included in the `bin`, `lib`, `lib64` set of directories, or the rpath derived from dependent packages. Consider the case where we have directory structure 

|---- bin --- someexe.exe
|
|-----lib
|                   | --- some-different-sub-project --- otherlib.dll
|-----other- |

where someexe.exe requires a link to otherlib.dll. Parsing would determine the link is internal and then link the bin to the other directory, whereas we want the other way around (the otherlib.dll linked to the bin/lib). 

Alternatively, we could forgo the internal links method and have an "add to link manifest" method of some sort where we include the directory of the desired binary/library that also requires linking against. If we add the packages bin to the rpath property definition (for windows only), this would resolve our clingo problem, and perhaps clarify usage?

Reply:
Ok so to make sure:

* Internal links is for registering directories/binaries/libraries that need to locate other libraries (could it in fact be any of these three or is it e.g. never directories?)
* `add_search_path`/`include_additional_link_paths` is for adding directories that the former would want to rpath to

if so, we should probably rename those here:

* `add_library_dependent`: for things that need to find libraries
* `add_library_dir` or `add_rpath`: for locations which include libs (`add_search_path` is also fine, this name is less confusing than `add_internal_links`)

Furthermore, instead of having a custom Package function that returns just the dependent paths (as I initially suggested in this specific thread), it can return two lists: the list of dependent paths, and the list of search paths. Would that cover it?